src/optimize.jl

```diff
@@ -0,0 +1,195 @@
+
+"""
+    optimize(
+        problem,
+        objective   ::AbstractVariationalObjective,
+        restructure,
+        param_init  ::AbstractVector{<:Real},
+        max_iter    ::Int,
+        objargs...;
+        kwargs...
+    )              
+    optimize(
+        problem,
+        objective             ::AbstractVariationalObjective,
+        variational_dist_init,
+        max_iter              ::Int,
+        objargs...;
+        kwargs...
+    )              
+
+Optimize the variational objective `objective` targeting the problem `problem` by estimating (stochastic) gradients, where the variational approximation can be constructed by passing the variational parameters `param_init` or the initial variational approximation `variational_dist_init` to the function `restructure`.
+
+# Arguments
+- `objective`: Variational Objective.
+- `param_init`: Initial value of the variational parameters.
+- `restruct`: Function that reconstructs the variational approximation from the flattened parameters.
+- `variational_dist_init`: Initial variational distribution. The variational parameters must be extractable through `Optimisers.destructure`.
+- `max_iter`: Maximum number of iterations.
+- `objargs...`: Arguments to be passed to `objective`.
+
+# Keyword Arguments
+- `adbackend::ADtypes.AbstractADType`: Automatic differentiation backend. 
+- `optimizer::Optimisers.AbstractRule`: Optimizer used for inference. (Default: `Adam`.)
+- `rng::AbstractRNG`: Random number generator. (Default: `Random.default_rng()`.)
+- `show_progress::Bool`: Whether to show the progress bar. (Default: `true`.)
+- `callback!`: Callback function called after every iteration. See further information below. (Default: `nothing`.)
+- `prog`: Progress bar configuration. (Default: `ProgressMeter.Progress(n_max_iter; desc="Optimizing", barlen=31, showspeed=true, enabled=prog)`.)
+- `state::NamedTuple`: Initial value for the internal state of optimization. Used to warm-start from the state of a previous run. (See the returned values below.)
+
+# Returns
+- `params`: Variational parameters optimizing the variational objective.
+- `stats`: Statistics gathered during optimization.
+- `state`: Collection of the final internal states of optimization. This can used later to warm-start from the last iteration of the corresponding run.
+
+# Callback
+The callback function `callback!` has a signature of
+
+    cb(; stat, state, param, restructure, gradient)
```

Comment:
```suggestion
    callback!(; stat, state, param, restructure, gradient)
```

Also, is there a scenario where `callback` is expected to be mutating?

By using the `!` notation, I'm expecting two things from this:
1. It mutates at least one of the inputs.
2. There is no guarantee that it's returning anything, unless explicitly stated so.

None of these properties seem true in our case, so we should probably just drop `!`, no?

Reply:
Agreed.